Reboot to Firmware Setup

meisenzahl opened the pull request:
If I ever need to make changes on the firmware (BIOS/UEFI) of my device, I often don't know which key combination to press when starting the device to get there. Windows offers an option to boot directly into BIOS/UEFI. On Linux, I have used https://www.freedesktop.org/software/systemd/man/systemctl.html#--firmware-setup for this so far. This feature adds a button that reboots the device and starts the BIOS/UEFI.

![Bildschirmfoto von 2022-10-30 05 55 54](https://user-images.githubusercontent.com/10796736/198876003-b3f7dc65-0fc5-4390-8199-bf57c17ccac9.png)

![Bildschirmfoto von 2022-10-30 05 56 11](https://user-images.githubusercontent.com/10796736/198876007-f611617f-8481-4866-ac4a-4c3f9876ed80.png)

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -125,17 +125,24 @@ public class About.OperatingSystemView : Gtk.Grid {
 
         var settings_restore_button = new Gtk.Button.with_label (_("Restore Default Settings"));
 
+        var reboot_to_firmware_setup_button = new Gtk.Button.with_label (_("Reboot to Firmware Setup"));
```
It looks like in the session indicator we're using "Restart" instead of "Reboot"
```suggestion
        var reboot_to_firmware_setup_button = new Gtk.Button.with_label (_("Restart to Firmware Setup"));
```

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -125,17 +125,24 @@ public class About.OperatingSystemView : Gtk.Grid {
         button_grid.add (settings_restore_button);
+        button_grid.set_child_secondary (settings_restore_button, true);
+        if (LoginManager.get_instance ().can_reboot_to_firmware_setup ()) {
+            button_grid.add (reboot_to_firmware_setup_button);
```
Can we add this on the Firmware page instead of the Operating System page? It seems like that's where it's most relevant

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -248,6 +257,33 @@ public class About.OperatingSystemView : Gtk.Grid {
 
+    private bool confirm_reboot_to_firmware_setup_action () {
+        var dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("Are you sure you want to Reboot?"),
+            _("This will close all open applications and turn off this device."),
```
We can probably be more specific here
```suggestion
            _("Restart to firmware setup?"),
            _("This will close all open applications, restart this device, and open the firmware setup screen."),
```

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -248,6 +257,33 @@ public class About.OperatingSystemView : Gtk.Grid {
+        var dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("Are you sure you want to Reboot?"),
+            _("This will close all open applications and turn off this device."),
+            "system-shutdown",
```
There's a more specific fd.o icon name for this:
```suggestion
            "system-reboot",
```

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -248,6 +257,33 @@ public class About.OperatingSystemView : Gtk.Grid {
+        );
+        dialog.transient_for = (Gtk.Window) get_toplevel ();
+
+        var continue_button = dialog.add_button (_("Reboot"), Gtk.ResponseType.ACCEPT);
```
Let's use Restart to be consistent
```suggestion
        var continue_button = dialog.add_button (_("Restart"), Gtk.ResponseType.ACCEPT);
```

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -248,6 +257,33 @@ public class About.OperatingSystemView : Gtk.Grid {
+        var continue_button = dialog.add_button (_("Reboot"), Gtk.ResponseType.ACCEPT);
+        continue_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+
+        var result = dialog.run ();
+        dialog.destroy ();
+
+        return result == Gtk.ResponseType.ACCEPT;
```
`dialog.run ()` no longer exists in Gtk 4, so we shouldn't use it in new code. Instead we should call `dialog.present ()` and then connect to `dialog.response ()`

danirabbit commented on src/Views/OperatingSystemView.vala:
```diff
@@ -248,6 +257,33 @@ public class About.OperatingSystemView : Gtk.Grid {
+    }
+
+    private void reboot_to_firmware_setup_clicked () {
+        if (confirm_reboot_to_firmware_setup_action ()) {
```
Kind of going with the other thing, we should connect to `dialog.response ()` since we can't get a return value from `run ()` anymore. This is kind of weird to have as a single-use function anyways

danirabbit commented on src/Interfaces/LoginManager.vala:
```diff
@@ -68,4 +70,22 @@ public class About.LoginManager : Object {
+        try {
+            interface.set_reboot_to_firmware_setup (true);
+        } catch (Error e) {
+            warning ("Could not connect to login interface: %s", e.message);
```
It seems like we don't want to fail silently here. If we can't set this flag then we should abort the system restart and throw an error to the user that we're unable to restart into the firmware setup screen. Otherwise we're restarting them for no reason with no explanation

danirabbit reviewed (changes requested):
Nice feature! I have a few inline comments

danirabbit commented:
I can confirm the feature works as expected, so happy to approve after the above is resolved :)

meisenzahl commented:
@danirabbit thank you very much for your feedback! I hope that I addressed all issues.

![Bildschirmfoto von 2022-11-01 18 45 04](https://user-images.githubusercontent.com/10796736/199309814-9eac78d6-f21d-47be-aec2-48a09a076a7a.png)

![Bildschirmfoto von 2022-11-01 18 53 30](https://user-images.githubusercontent.com/10796736/199310149-a9dd422f-b5b4-47a8-be87-b7eeb48928dc.png)

![Bildschirmfoto von 2022-11-01 19 27 42](https://user-images.githubusercontent.com/10796736/199310178-c550de13-f4d3-4be6-afce-3f7103c2572d.png)

I was not sure if there is already an established way of handling dialogs as preparation for porting to Gtk4. Also feel free to improve the error dialog texts.

danirabbit commented:
@meisenzahl made a branch here :) https://github.com/elementary/switchboard-plug-about/pull/254

danirabbit reviewed (approved):
Let's go! Nice new feature :rocket:
